### ISSUE TITLE
"Auto-registration of the command..." deprecation warnings thrown during testing for no discernible reason

### ISSUE THREAD

**Issue by arderyp** [symfony/symfony#28121]
**Symfony version(s) affected**: 3.4.11

**Description**  

PHPUnit (`6.5.8`) via PHPUnitBridge (`4.1.0`) is throwing **Auto-registration of the command...** warnings for no discernible reason for `DoctrineMigrations` commands when testing my own custom commands.  My custom commands have nothing to do with `DoctrineMigrations`.  Furthermore, I've looked at the `DoctrineMigrations` code, and it appears that the commands are configured properly.

**How to reproduce**  

Include `DoctrineMigrations` in your project, create your own custom command, and write a [unit test](https://symfony.com/doc/current/console.html#testing-commands) for your custom command.

**Possible Solution**  

The warning is being thrown by [this line](https://github.com/symfony/http-kernel/blob/3.4/Bundle/Bundle.php#L189) in `Bundle.php`.  `Doctrine\Bundle\FixturesBundle\Command\LoadDataFixturesDoctrineCommand`, for example, doesn't throw this warning, but all of the `DoctrineMigration` commands do because of the third qualifier in the line I linked above:

```
!$r->getConstructor()->getNumberOfRequiredParameters()
```

All `DoctrineMigration` command fulfill that, while other commands like `Doctrine\Bundle\FixturesBundle\Command\LoadDataFixturesDoctrineCommand` don't.  Can someone explain why the number of required constructor arguments for a given command is even a qualifier for throwing the deprecation warning?

**Additional context**  

I originally posted this information on a `DoctrineMigrations` issue [here](https://github.com/doctrine/DoctrineMigrationsBundle/issues/221), but this seems more and more like an issue with core, or an issue that I am simply not understanding properly.  Any insight is greatly appreciated.

**Comment by arderyp** [comment]
Hmm, if I get the time to, I can try.  Do you (or anyone else) by chance know why this qualifier is being used?

```
!$r->getConstructor()->getNumberOfRequiredParameters()
```

**Comment by xabbuh** [comment]
Command classes with mandatory constructor arguments were never registered implicitly. So there is no need to trigger a deprecation for them.

**Comment by arderyp** [comment]
And those without mandatory constructor arguments are always invalid in all cases, even if they are configured correctly according to current auto-wiring standards?

**Comment by xabbuh** [comment]
When they are correctly registered as services, you shouldn't see the deprecation. If that's not the case, we will need a reproducible example.

**Comment by arderyp** [comment]
So you're saying that a command with no constructor arguments that's correctly registered shouldn't fall into:

```
!$r->getConstructor()->getNumberOfRequiredParameters()
```

There's something about registration that causes this core PHP function `getNumberOfRequiredParameters` to behave differently?

**Comment by arderyp** [comment]
Well, the more I dig into this, the more confused I get.  I am not able to reproduce the issue in a simple Symfony skeleton.  The real infrastructure it is being thrown in is rather complex.  From what I can gather, the issue it not stemming from the `DoctrineMigrations` command/service configuration.  Thus, I can only assume that the problem lies either in my approach to testing console commands (I've done a lot of testing and am following standard Symfony documentation), or in some obscure use-case issue in core that my code is falling into.

I do not see this error elsewhere in the logs or browser profiler tool, and the `DoctrineMigration` commands in question work just fine and don't throw warnings/error.

It doesn't seem that anyone is able to explain the mysterious (uncommented) use of the below qualifier, which is triggering the error:

```
!$r->getConstructor()->getNumberOfRequiredParameters()
```

The root of the problem is that the `add()` method on `Application` first tries to register all commands via the `registerCommands()` method before adding the command in question.  This is yet another point of confusion for me.  Why is this happening?  It seems like a weird design pattern that every method on `Application` tries to run `registerCommands()` before doing any work.  I don't think I understand the full workflow properly.

The reality is that, in our example of testing a single custom command, we don't need or want to register all of the system commands.  We simply want to test our single custom command, which we are registering explicitly through `add()`.  Because of this, I've come up with a workaround that might generally be ill advised, but it seems perfectly reasonable to me given the reasoning in the sentence preceding to this one.

Ultimately, it seems wrong that I should have to take the approach below.  That being said, I'll have to stick to it for the time being; so, if anyone notices some inherent problem with this workaround, please do let me know.

## Workaround
instead of doing this:

```
$mockCommand = ...;
$application = new Application(self::bootKernel());
$application->add($mockCommand);
```

mock the `Application` object and suppress the problematic `registerCommands()` method as follows:

```
$mockCommand = ...;

// mock Application object to prevent registering all console commands
// since we only want to register and test our single custom command
$constructorArgs = [self::bootKernel()];
$methodRegisterCommands = 'registerCommands';
$mockApplication = $this->getMockBuilder(Application::class)
    ->setConstructorArgs($constructorArgs)
    ->setMethods([$methodRegisterCommands])
    ->getMock();
$mockApplication->method($methodRegisterCommands)->willReturn(null);

$application = $mockApplication;
$application->add($mockCommand);
```

**Comment by xabbuh** [comment]
As I wrote before this line is there to only automatically register commands that do not require mandatory constructor arguments. The application does not know what to pass for these arguments and thus always used to slip these command classes when registering commands provided by bundles.

Why that does not work for you is hard to say without having an example that allows to reproduce your issue.

**Comment by arderyp** [comment]
It looks like the `console.command.ids` parameter is not being hydrated properly in my system.  Again, I suspect this is not a Symfony bug, but some problem with my system.configuration.  I'd greatly appreciate any feetback in that regard.

On my system, if I dump `$this->container->getParameter('console.command.ids')` from `Bundle.php`, I get the following:

```
array(17) {
  [0]=>
  string(70) "console.command.public_alias.Path\To\My\Custom\Command\FirstCommand"
  [1]=>
  string(71) "console.command.public_alias.Path\To\My\Custom\Command\SecondCommand"
  [2]=>
  string(74) "console.command.public_alias.Path\To\My\Custom\Command\ThirdCommand"
  [4]=>
  string(60) "console.command.public_alias.doctrine_cache.contains_command"
  [5]=>
  string(58) "console.command.public_alias.doctrine_cache.delete_command"
  [6]=>
  string(57) "console.command.public_alias.doctrine_cache.flush_command"
  [7]=>
  string(57) "console.command.public_alias.doctrine_cache.stats_command"
  [8]=>
  string(61) "console.command.public_alias.doctrine_migrations.diff_command"
  [9]=>
  string(64) "console.command.public_alias.doctrine_migrations.execute_command"
  [10]=>
  string(65) "console.command.public_alias.doctrine_migrations.generate_command"
  [11]=>
  string(63) "console.command.public_alias.doctrine_migrations.latest_command"
  [12]=>
  string(64) "console.command.public_alias.doctrine_migrations.migrate_command"
  [13]=>
  string(63) "console.command.public_alias.doctrine_migrations.status_command"
  [14]=>
  string(64) "console.command.public_alias.doctrine_migrations.version_command"
}
```

If I dump the same parameter in a fresh Symfony 3.4 skeleton with (more or less) the same composer dependency environment, I get the following:

```
array(60) {
  ["console.command.appbundle_command_customcommand"]=>
  string(47) "console.command.appbundle_command_customcommand"
  ["console.command.symfony_bundle_frameworkbundle_command_aboutcommand"]=>
  string(21) "console.command.about"
  ["console.command.symfony_bundle_frameworkbundle_command_assetsinstallcommand"]=>
  string(30) "console.command.assets_install"
  ["console.command.symfony_bundle_frameworkbundle_command_cacheclearcommand"]=>
  string(27) "console.command.cache_clear"
  ["console.command.symfony_bundle_frameworkbundle_command_cachepoolclearcommand"]=>
  string(32) "console.command.cache_pool_clear"
  ["console.command.symfony_bundle_frameworkbundle_command_cachepoolprunecommand"]=>
  string(32) "console.command.cache_pool_prune"
  ["console.command.symfony_bundle_frameworkbundle_command_cachewarmupcommand"]=>
  string(28) "console.command.cache_warmup"
  ["console.command.symfony_bundle_frameworkbundle_command_configdebugcommand"]=>
  string(28) "console.command.config_debug"
  ["console.command.symfony_bundle_frameworkbundle_command_configdumpreferencecommand"]=>
  string(37) "console.command.config_dump_reference"
  ["console.command.symfony_bundle_frameworkbundle_command_containerdebugcommand"]=>
  string(31) "console.command.container_debug"
  ["console.command.symfony_bundle_frameworkbundle_command_debugautowiringcommand"]=>
  string(32) "console.command.debug_autowiring"
  ["console.command.symfony_bundle_frameworkbundle_command_eventdispatcherdebugcommand"]=>
  string(38) "console.command.event_dispatcher_debug"
  ["console.command.symfony_bundle_frameworkbundle_command_routerdebugcommand"]=>
  string(28) "console.command.router_debug"
  ["console.command.symfony_bundle_frameworkbundle_command_routermatchcommand"]=>
  string(28) "console.command.router_match"
  ["console.command.symfony_bundle_frameworkbundle_command_xlifflintcommand"]=>
  string(26) "console.command.xliff_lint"
  ["console.command.symfony_bundle_frameworkbundle_command_yamllintcommand"]=>
  string(25) "console.command.yaml_lint"
  ["console.command.symfony_component_form_command_debugcommand"]=>
  string(26) "console.command.form_debug"
  ["console.command.symfony_bundle_securitybundle_command_userpasswordencodercommand"]=>
  string(38) "security.command.user_password_encoder"
  ["console.command.symfony_bridge_twig_command_debugcommand"]=>
  string(18) "twig.command.debug"
  ["console.command.symfony_bundle_twigbundle_command_lintcommand"]=>
  string(17) "twig.command.lint"
  ["console.command.symfony_bundle_swiftmailerbundle_command_debugcommand"]=>
  string(25) "swiftmailer.command.debug"
  ["console.command.symfony_bundle_swiftmailerbundle_command_newemailcommand"]=>
  string(29) "swiftmailer.command.new_email"
  ["console.command.symfony_bundle_swiftmailerbundle_command_sendemailcommand"]=>
  string(30) "swiftmailer.command.send_email"
  ["console.command.doctrine_bundle_doctrinecachebundle_command_containscommand"]=>
  string(75) "console.command.doctrine_bundle_doctrinecachebundle_command_containscommand"
  ["console.command.doctrine_bundle_doctrinecachebundle_command_deletecommand"]=>
  string(73) "console.command.doctrine_bundle_doctrinecachebundle_command_deletecommand"
  ["console.command.doctrine_bundle_doctrinecachebundle_command_flushcommand"]=>
  string(72) "console.command.doctrine_bundle_doctrinecachebundle_command_flushcommand"
  ["console.command.doctrine_bundle_doctrinecachebundle_command_statscommand"]=>
  string(72) "console.command.doctrine_bundle_doctrinecachebundle_command_statscommand"
  ["console.command.doctrine_bundle_doctrinebundle_command_createdatabasedoctrinecommand"]=>
  string(32) "doctrine.database_create_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_dropdatabasedoctrinecommand"]=>
  string(30) "doctrine.database_drop_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_generateentitiesdoctrinecommand"]=>
  string(34) "doctrine.generate_entities_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_importmappingdoctrinecommand"]=>
  string(31) "doctrine.mapping_import_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_clearmetadatacachedoctrinecommand"]=>
  string(37) "doctrine.cache_clear_metadata_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_clearquerycachedoctrinecommand"]=>
  string(40) "doctrine.cache_clear_query_cache_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_clearresultcachedoctrinecommand"]=>
  string(35) "doctrine.cache_clear_result_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_collectionregiondoctrinecommand"]=>
  string(40) "doctrine.cache_collection_region_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_convertmappingdoctrinecommand"]=>
  string(32) "doctrine.mapping_convert_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_createschemadoctrinecommand"]=>
  string(30) "doctrine.schema_create_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_dropschemadoctrinecommand"]=>
  string(28) "doctrine.schema_drop_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_ensureproductionsettingsdoctrinecommand"]=>
  string(43) "doctrine.ensure_production_settings_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_entityregioncachedoctrinecommand"]=>
  string(36) "doctrine.clear_entity_region_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_importdoctrinecommand"]=>
  string(32) "doctrine.database_import_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_infodoctrinecommand"]=>
  string(29) "doctrine.mapping_info_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_queryregioncachedoctrinecommand"]=>
  string(35) "doctrine.clear_query_region_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_rundqldoctrinecommand"]=>
  string(26) "doctrine.query_dql_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_runsqldoctrinecommand"]=>
  string(26) "doctrine.query_sql_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_updateschemadoctrinecommand"]=>
  string(30) "doctrine.schema_update_command"
  ["console.command.doctrine_bundle_doctrinebundle_command_proxy_validateschemacommand"]=>
  string(32) "doctrine.schema_validate_command"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationsdiffdoctrinecommand"]=>
  string(86) "console.command.doctrine_bundle_migrationsbundle_command_migrationsdiffdoctrinecommand"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationsexecutedoctrinecommand"]=>
  string(89) "console.command.doctrine_bundle_migrationsbundle_command_migrationsexecutedoctrinecommand"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationsgeneratedoctrinecommand"]=>
  string(90) "console.command.doctrine_bundle_migrationsbundle_command_migrationsgeneratedoctrinecommand"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationslatestdoctrinecommand"]=>
  string(88) "console.command.doctrine_bundle_migrationsbundle_command_migrationslatestdoctrinecommand"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationsmigratedoctrinecommand"]=>
  string(89) "console.command.doctrine_bundle_migrationsbundle_command_migrationsmigratedoctrinecommand"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationsstatusdoctrinecommand"]=>
  string(88) "console.command.doctrine_bundle_migrationsbundle_command_migrationsstatusdoctrinecommand"
  ["console.command.doctrine_bundle_migrationsbundle_command_migrationsversiondoctrinecommand"]=>
  string(89) "console.command.doctrine_bundle_migrationsbundle_command_migrationsversiondoctrinecommand"
  ["console.command.sensiolabs_security_command_securitycheckercommand"]=>
  string(44) "sensio_distribution.security_checker.command"
  ["console.command.symfony_bundle_webserverbundle_command_serverruncommand"]=>
  string(29) "web_server.command.server_run"
  ["console.command.symfony_bundle_webserverbundle_command_serverstartcommand"]=>
  string(31) "web_server.command.server_start"
  ["console.command.symfony_bundle_webserverbundle_command_serverstopcommand"]=>
  string(30) "web_server.command.server_stop"
  ["console.command.symfony_bundle_webserverbundle_command_serverstatuscommand"]=>
  string(32) "web_server.command.server_status"
  ["console.command.symfony_bundle_webserverbundle_command_serverlogcommand"]=>
  string(29) "web_server.command.server_log"
}
```

I wonder what I am missing in my main system/configuration that could be causing `console.command.ids` to be incomplete/malformed (missing vendor commands and keys as integers instead of the strings defined in config files).  Anyone have any ideas?  The one major difference in the two environments above is that my system is using FLEX, while the fresh Symfony skeleton is not.

**Comment by arderyp** [comment]
Right, I know @xabbuh, but my example application (second code block above) is not problematic.  It works properly and doesn't throw the dependency error (I couldn't reproduce the issue there).  So, I don't see any point in my pushing and linking a sample project that works properly.  My actual application is not open sourced at the moment, and I cannot share it here.  I know that's not helpful, but I'm doing my best to simply solicit advice from users more knowledgeable about Symfony core than I am.  As I mentioned, I don't think this is a bug.... its a problem with my system that hopefully someone here can help me resolve.

**Comment by xabbuh** [comment]
Okay, in this case I am going to close here then. We can reopen anytime in the future if we discover that this really is a bug and we know how to reproduce it.

**Comment by arderyp** [comment]
Thanks for your patience @xabbuh.  Coincidentally, I just figured out the problem (for anyone else experiencing this problem).  In moving to Symfony FLEX, my `composer.json` was no longer including `symfony/symfony`, but rather the explicit symfony components/packages (as outlined [here](https://symfony.com/doc/current/setup/flex.html#upgrading-existing-applications-to-flex)). As it turns out, I did not explicitly require `"symfony/console": "^3.4",`, which was causing version`4.1` of this package to be installed (as it is required by some other symfony component, and was defaulting to the latest version, I suppose).  I wasn't explicitly requiring various other packages as well, so I think my system was running in kind of a half 3.4 half 4.1 state.  Once I fixed my `composer.json`, wiped out `vendor/` and re-ran `composer install`, the warnings went away.
